components/movementsensor/merged/merged.go

```diff
@@ -0,0 +1,373 @@
+			// names the same as the sesnor currently used by this model
+			return (cfgVals[0] != ms.Name().ShortName())
+		}
+
+		return false
+	}
+
+	oriChanged := sensorChanged(newConf.Orientation, m.ori)
+	posChanged := sensorChanged(newConf.Position, m.pos)
+	compassChanged := sensorChanged(newConf.CompassHeading, m.compass)
+	lvelChanged := sensorChanged(newConf.LinearVelocity, m.linVel)
+	avelChanged := sensorChanged(newConf.AngularVelocity, m.angVel)
+	laccChanged := sensorChanged(newConf.LinearAcceleration, m.linAcc)
+
+	firstGoodSensorWithProperties := func(deps resource.Dependencies, names []string, mprops *movementsensor.Properties,
+	) (movementsensor.MovementSensor, error) {
+		// check if the config names and dependencies have been passed at all
+		if len(names) == 0 || deps == nil {
+			return nil, nil
+		}
+
+		var ms movementsensor.MovementSensor
+		for _, name := range names {
+			depsms, err := movementsensor.FromDependencies(deps, name)
+			if err == nil {
+				ms = depsms
+				// found the first good sensor for this information
+				// don't check the rest
+				break
+			} else {
+				continue
+			}
+		}
+
+		props, err := ms.Properties(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		if reflect.DeepEqual(props, mprops) {
+			return ms, nil
+		}
+		return nil, fmt.Errorf("none of the sensors in provided list %#v statisfy properties, need %#v", names, mprops)
+	}
+
+	if oriChanged {
+		m.ori, err = firstGoodSensorWithProperties(deps,
+			newConf.Orientation,
+			&movementsensor.Properties{OrientationSupported: true})
+		if err != nil {
+			return err
+		}
+		m.oriReporting = true
+	}
+
+	if posChanged {
+		m.pos, err = firstGoodSensorWithProperties(deps,
+			newConf.Position,
+			&movementsensor.Properties{PositionSupported: true})
+		if err != nil {
+			return err
+		}
+		m.posReporting = true
+	}
+
+	if compassChanged {
+		m.compass, err = firstGoodSensorWithProperties(deps,
+			newConf.CompassHeading,
+			&movementsensor.Properties{CompassHeadingSupported: true})
+		if err != nil {
+			return err
+		}
+		m.compassReporting = true
+	}
+
+	if lvelChanged {
+		m.linVel, err = firstGoodSensorWithProperties(deps,
+			newConf.LinearVelocity,
+			&movementsensor.Properties{LinearVelocitySupported: true})
+		if err != nil {
+			return err
+		}
+		m.linVelReporting = true
+	}
+
+	if avelChanged {
+		m.angVel, err = firstGoodSensorWithProperties(deps,
+			newConf.AngularVelocity,
+			&movementsensor.Properties{AngularVelocitySupported: true})
+		if err != nil {
+			return err
+		}
+		m.angVelReporting = true
+	}
+
+	if laccChanged {
+		m.linAcc, err = firstGoodSensorWithProperties(deps,
+			newConf.LinearAcceleration,
+			&movementsensor.Properties{LinearAccelerationSupported: true})
+		if err != nil {
+			return err
+		}
+		m.linAccReporting = true
+	}
+
+	return nil
+}
+
+func (m *merged) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	if m.pos == nil {
```

Comment:
I'm worried that the underlying sensor will have a lock, should I do:
```go
mu.Lock()
if m.pos == nil {
handleNilSensor()
}
out, err := m.pos.Position(ctx, nil)
handleError()
mu.Unlock()
return out, nil
```

Reply:
Assuming that every function which locks a mutex unlocks it before the function returns, the two locks won't interfere with each other. The code I'm thinking of is this:
```go
func (m *merged) Position(...) {
    m.mu.Lock()
    defer m.mu.Unlock()

    <everything else exactly as you've already written it>
}
```
Here's the situation to avoid:
1. We've got this component already configured with a position sensor
2. Thread A calls `Position()`. It sees that `m.pos` is not nil, and continues past this if statement.
3. Thread B calls `Reconfigure()`. It locks the mutex. In the new config, there is no position sensor, and it sets `m.pos` to nil.
4. Thread A calls `m.pos.Position()`, and runs into trouble because there's a nil value in an unexpected place.
5. Thread B finishes reconfiguring and unlocks the mutex again.

By locking the mutex at the top of `Position` (and the other getter functions), we guarantee that `Position()` and `Reconfigure()` cannot run at the same time, and we don't need to worry about this sort of race condition. By locking the mutex in step 2, we force step 4 to run before step 3, which is how we avoid the trouble.